### DISCUSSION TITLE
State history requiring a query and not a store

### DISCUSSION THREAD

**Discussion by milanov** [salesforce/akita#751] [Q&A]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[ ] Feature request
[x] Documentation issue or request
[ ] Support request
[ ] Other... Please describe:
</code></pre>

## Current behavior

We introduced Akita in our projects recently. The store and query abstractions provide a good way to encapsulate their respective functionalities. With that in mind, we wanted to also encapsulate the undo/redo. It felt natural to expose them as methods of the various `Store`s, as the documentation [states](https://datorama.github.io/akita/docs/plugins/state-history/) that:

> From the moment you call it, Akita's StateHistory tracks the history of the **store**

I was surprised, though, as the state history plugin requires a query. We ended up with methods like `EntityQuery#undo(entityId)`, which feels out of place, as I think queries are supposed to only query.

## Expected behavior

Could you elaborate (or document) why queries are needed for the state history, and not directly a store?

**Comment by NetanelBasal** [comment]
Because we need to listen to state changes:

https://github.com/datorama/akita/blob/master/libs/akita/src/lib/plugins/stateHistory/stateHistoryPlugin.ts#L91
